app/src/main/java/com/github/h3lp3rs/h3lp/forum/FireForum.kt

```diff
@@ -0,0 +1,262 @@
+                    }
+                }
+                // We add the listener on the POSTS_LIST key since as explained in the protocol,
+                // it is updated with new post keys on every new post
+                rootForum.addEventListener(
+                    pathToKey(path + POSTS_LIST),
+                    String::class.java,
+                    onNewPost
+                ) {}
+            }
+            else -> {
+                // Listening to a single post means listening on all of its children
+                getPost(emptyList()).thenAccept { forumPost ->
+                    val repliesPath = forumPost.forum.path + forumPost.post.repliesKey
+                    // Adding a listener on the replies key since this is where all replies to that
+                    // post are stored
+                    rootForum.addEventListener(
+                        pathToKey(repliesPath),
+                        ForumPostData::class.java,
+                        action
+                    ) {}
+                }
+            }
+        }
+    }
+
+    override fun root(): Forum {
+        return FireForum(emptyList())
+    }
+
+    override fun child(relativePath: Path): Forum {
+        return FireForum(path + relativePath)
+    }
+
+    override fun parent(): Forum {
+        return if (isRoot()) {
+            this
+        } else {
+            FireForum(path.dropLast(1))
+        }
+    }
+
+    /**
+     * Translates a Path into its corresponding key as understood by Firebase (in Firebase, strings
+     * separated by a "/" represent a path, as in a filesystem with "files" or children in the case
+     * of a database)
+     * @param path The path to transform into a string according to Firebase
+     * @return The corresponding string
+     */
+    private fun pathToKey(path: Path): String {
+        return path.joinToString(separator = "/")
+    }
+
+    /**
+     * Abstracts away the implementation of the path into the Forum structure, here, the root is
+     * represented by an empty path
+     * @return A boolean corresponding to the fact that this forum is (or isn't) the root forum
+     */
+    private fun isRoot(): Boolean {
+        return path.isEmpty()
+    }
+
+
+    /**
+     * Abstracts away the implementation of the path into the Forum structure, here, a category is
+     * represented by a path of length 1
+     * @return A boolean corresponding to the fact that this forum is (or isn't) the root forum
+     */
+    private fun isCategory(): Boolean {
+        return path.size == 1
+    }
+
+    /**
+     * Returns the category of the current forum and the default category "general" in case the
+     * current forum is root
+     */
+    private fun getCurrentCategory(): ForumCategory {
+        return if (!isRoot()) {
+            valueOf(path.first())
+        } else {
+            // The default forum category is "general"
+            GENERAL
```

Comment:
Would be nice to put this in a general const variable like DEFAULT_CATEGORY in case it changes in the future and if some other classes need to know the default value for the forum category.